blog post: Istio as a proxy for external services (#3180)
istio/istio.io: content/docs/examples/advanced-gateways/front-proxy/index.md

---
title: Istio as a Front Proxy for External Services
description: Describes how to configure ingress gateway as a front proxy for external services.
weight: 9
keywords: [traffic-management,ingress,https,http]
---

> The [Control Ingress Traffic](/docs/tasks/traffic-management/ingress) task and the

Author: Addressed by https://github.com/istio/istio.io/pull/3180/commits/f5e43314f83ca6b091479607061229fac3b10bd5

      servers:
      - port:
          number: 80
          name: http
          protocol: HTTP
        hosts:
        - httpbin.org
      - port:
          number: 443
          name: https
          protocol: HTTPS
        tls:
          mode: PASSTHROUGH
        hosts:
        - www.google.com
    EOF
    {{< /text >}}

1.  Create service entries for `httpbin.org` and `www.google.com` to make them accessible from the ingress gateway:

    {{< text bash >}}
    $ kubectl apply -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: ServiceEntry
    metadata:
      name: httpbin-ext
    spec:
      hosts:
      - httpbin.org
      ports:
      - number: 80
        name: http
        protocol: HTTP
      resolution: DNS
      location: MESH_EXTERNAL
    ---
    apiVersion: networking.istio.io/v1alpha3
    kind: ServiceEntry
    metadata:
      name: google
    spec:
      hosts:
      - www.google.com
      ports:
      - number: 443
        name: https
        protocol: HTTPS
      resolution: DNS
      location: MESH_EXTERNAL
    EOF
    {{< /text >}}

1.  Configure a service entry and a destination rule for `localhost`. This service entry is used in the next step

>     as a destination for traffic to the external services from applications inside the mesh, effectively blocking such

Author: Addressed by https://github.com/istio/istio.io/pull/3180/commits/52abca262d0391f72828ab07117889007cf9e740

    traffic.

    {{< text bash >}}
    $ kubectl apply -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: ServiceEntry
    metadata:
      name: localhost
    spec:
      hosts:
      - localhost.local
      location: MESH_EXTERNAL
      ports:
      - number: 80
        name: http
        protocol: HTTP
      - number: 443
        name: https
        protocol: HTTPS
      resolution: STATIC
      endpoints:
      - address: 127.0.0.1
    ---
    apiVersion: networking.istio.io/v1alpha3
    kind: DestinationRule
    metadata:
      name: localhost
    spec:
      host: localhost.local
      trafficPolicy:
        tls:
          mode: DISABLE
          sni: localhost.local
    EOF
    {{< /text >}}

> 1.  Configure routing. For pods inside the mesh, direct the traffic to `localhost.local`, effectively blocking it.

Author: Addressed by https://github.com/istio/istio.io/pull/3180/commits/03392636a187ae359d02ce1b53c289aa9de02d90

            port:
              number: 80
      - match:
        - gateways:
          - mesh
          port: 80
        route:
        - destination:
            host: localhost.local
            port:
              number: 80
    ---
    apiVersion: networking.istio.io/v1alpha3
    kind: VirtualService
    metadata:
      name: google
    spec:
      hosts:
      - www.google.com
      gateways:
      - front-proxy
      - mesh
      tls:
      - match:
        - gateways:
          - front-proxy
          port: 443
          sni_hosts:
          - www.google.com
        route:
        - destination:
            host: www.google.com
            port:
              number: 443
      - match:
        - gateways:
          - mesh
          port: 443
          sni_hosts:
          - www.google.com
        route:
        - destination:
            host: localhost.local
            port:
              number: 443
    EOF
    {{< /text >}}

1.  Follow the instructions in
    [Determining the ingress IP and ports](/docs/tasks/traffic-management/ingress/#determining-the-ingress-ip-and-ports)
    to define the `SECURE_INGRESS_PORT` and `INGRESS_HOST` environment variables.

> 1.  Access `httbin.org/status/418` through your ingress:

Author: Addressed by https://github.com/istio/istio.io/pull/3180/commits/8c30f2b6431678cf8e484cc899850b6a72e8b984

    {{< text bash >}}
    $ curl $INGRESS_HOST:$INGRESS_PORT/status/418 -Hhost:httpbin.org

    -=[ teapot ]=-

       _...._
     .'  _ _ `.
    | ."` ^ `". _,
    \_;`"---"`|//
      |       ;/
      \_     _/
        `"""`
    {{< /text >}}

> 1.  Check the logs of the `istio-ingressgateway` pods. If the gateway is deployed in the `istio-system` namespace,

Reviewer: ```suggestion
1.  If the Istio ingress gateway is deployed in the `istio-system` namespace, print the gateway's log with:
```

    the command to print the log is:

    {{< text bash >}}
    $ kubectl logs -l istio=ingressgateway -c istio-proxy -n istio-system | grep 'httpbin.org'
    {{< /text >}}

>     You should see a line similar to the following:

Reviewer: ```suggestion
1. Search the log for an entry similar to:
```

    {{< text plain >}}
    [2019-01-31T14:40:18.645Z] "GET /status/418 HTTP/1.1" 418 - 0 135 187 186 "10.127.220.75" "curl/7.54.0" "28255618-6ca5-9d91-9634-c562694a3625" "httpbin.org" "34.232.181.106:80" outbound|80||httpbin.org - 172.30.230.33:80 10.127.220.75:52077 -
    {{< /text >}}

> 1.  Check the Mixer log. If Istio is deployed in the `istio-system` namespace, the command to print the log is:

Reviewer: ```suggestion
1.  If Istio is deployed in the `istio-system` namespace, print the Mixer log with:
```

    {{< text bash >}}
    $ kubectl -n istio-system logs -l istio-mixer-type=telemetry -c mixer | grep 'httpbin.org'
    {{< /text >}}

>     You should see a line similar to the following:

Reviewer: ```suggestion
1. Search the log for an entry similar to:
```

    {{< /text >}}

> 1.  Access `www.google.com` through your ingress:

Reviewer: ```suggestion
1.  Access the `www.google.com` service through your ingress gateway:
```

    {{< text bash >}}
    $ curl -s --resolve www.google.com:$SECURE_INGRESS_PORT:$INGRESS_HOST https://www.google.com:$SECURE_INGRESS_PORT | grep -o "<title>.*</title>"
    <title>Google</title>
    {{< /text >}}

> 1.  Check the logs of the `istio-ingressgateway` pods. If the gateway is deployed in the `istio-system` namespace,

Reviewer: ```suggestion
1. If the Istio ingress gateway is deployed in the `istio-system` namespace, print the gateway's log with:
```

    the command to print the log is:

    {{< text bash >}}
    $ kubectl logs -l istio=ingressgateway -c istio-proxy -n istio-system | grep 'www.google.com'
    {{< /text >}}

>     You should see a line similar to the following:

Reviewer: ```suggestion
1. Search the log for an entry similar to:
```

    {{< text plain >}}
    [2019-01-31T13:40:11.076Z] "- - -" 0 - 589 17798 1644 - "-" "-" "-" "-" "172.217.31.132:443" outbound|443||www.google.com 172.30.230.33:54508 172.30.230.33:443 10.127.220.75:49467 www.google.com
    {{< /text >}}

> 1.  Check the Mixer log. If Istio is deployed in the `istio-system` namespace, the command to print the log is:

Reviewer: ```suggestion
1. If Istio is deployed in the `istio-system` namespace, print the Mixer log with:
```

    {{< text bash >}}
    $ kubectl -n istio-system logs -l istio-mixer-type=telemetry -c mixer | grep 'www.google.com' | grep '"sourceWorkload":"istio-ingressgateway"'| grep '"connectionEvent":"open"'
    {{< /text >}}

>     You should see a line similar to the following:

Reviewer: ```suggestion
1. Search the log for an entry similar to:
```
